### ISSUE TITLE
WCAG violation: drag-and-drop file upload element not accessible with keyboard

### ISSUE THREAD

**Issue by fvwanja** [solspace/craft3-freeform#2032]
### What happened?

Hey there 👋

The drag-and-drop file upload element (`[data-freeform-file-upload="fileDragAndDrop"]`) is not accessible with keyboard controls. In Chrome and Firefox on MacOS I can not focus it using the tab key. Chrome on Windows is sometimes okay, but seems inconsistent. MacOS also has an issue with Apple VoiceOver.

## Firefox

When using the tab-key or shift+tab-key, I can not interact with the file upload field. This is a violation against the WCAG rule 2.1.1. On June 28th or 2025, this violation becomes punishable by law in Germany ([BFSG](https://www.bmas.de/DE/Service/Gesetze-und-Gesetzesvorhaben/barrierefreiheitsstaerkungsgesetz.html)) – which will force users of the Freeform plugin to look for alternative software solutions to avoid fines.

https://github.com/user-attachments/assets/6cf33c62-36e0-4d30-bbac-4242bef8a8c8

I think this behaviour happens, because a generic `HTMLDivElement` is used – without the correct events and attributes attached. Replacing the generic element with an interactable `HTMLButtonElement` could help.

## VoiceOver

The element that is being focussed with VoiceOver on the first interaction is `div[data-placeholder].freeform-file-dnd__placeholder` – which contains the instruction on how to use the drag and drop field. However I would assume that I can interact with this element to upload a file – at least this is what the default label indicates: „Ziehen Sie die Dateien hierher oder klicken Sie zum Hochladen” (translated: „Drag the files here or click to upload”). The placeholder element should not be accessible at all.

![Image](https://github.com/user-attachments/assets/b1f7662f-8de5-44db-a873-8b18ed16076c)

Instead just reference it on the actual element, that handles the click event:

```html
<button aria-describedby="placeholder99">File upload</button>
<div id="placeholder99" aria-hidden="true">Drag the files here or click to upload</div>
```

The screenreader reads: "File upload, Drag the files here or click to upload" and the only interactable element is a real `HTMLButtonElement`. Yay 🥳

If you need any further information, feel free to ask.

Thank you for your time and effort 🙏

### Errors and Stack Trace (if available)

```shell

```

### How can we reproduce this?

1. Create a form with a drag-and-drop element
2. Visit the form
3. Use the tab-key or a screenreader and try to upload a file without using the mouse

### Freeform Edition

Pro

### Freeform Version

5.10.16.1

### Craft Version

Pro 4.15.6.2

### When did this issue start?

- [x] Unsure
- [ ] Fresh install of Freeform
- [ ] After upgrading from older Freeform version
- [ ] After upgrading from older Craft version
- [ ] After a change to hosting configuration

**Comment by kjmartens** [comment]
Hi @fvwanja,

Thanks for reporting this. I'll see if we can figure out a solution for this soon. 🙂

**Comment by kjmartens** [comment]
This should now be resolved in Freeform 5.12.2+ 🙂

**Comment by fvwanja** [comment]
Hello @kjmartens ,

thank you for the update. I can now access the field with the TAB Key in Firefox as I'd expect. I was also able to upload files with VoiceOver.

Unfortunately the HTML that is now generated is not valid. You cannot put a`<button>` inside a `<button>`. This will not only result in a list of HTML validation errors, but also makes the nested buttons inaccessible.

![Image](https://github.com/user-attachments/assets/0dc6f1c7-9df0-48e2-b078-81ac7153d76a)

In practice this means, that I am able to add files with a screenreader, but I cannot delete or read out files that I already picked, because of the nested HTML. I know you can do it with the TAB key, but a screenreader user would not expect a button within a button and it is also not being read out – so you have to know it is there, which makes it inaccessible.

![Image](https://github.com/user-attachments/assets/5cf1c2f4-8c04-48f2-be52-0939fb8ac2e1)

I suggest you build the HTML like this (simplified). This allows users to easily pick new files and edit the list of files.

```html
<form action="">
  <button type="button">Upload files</button>

  <ul aria-label="Uploaded files">
    <li>
      <span id="file-1">File 1</span>
      <button type="button" aria-label="Delete" aria-describedby="file-1">X</button>
    </li>
  </ul>
</form>
```

I'd like to repoen this issue. Thank you in advance!
